spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResourceBucket.java

```diff
@@ -0,0 +1,163 @@
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.storage;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.BucketInfo;
+import com.google.cloud.storage.Storage;
+
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.WritableResource;
+import org.springframework.util.Assert;
+
+/**
+ * Implements {@link WritableResource} for buckets in Google Cloud Storage.
+ */
+public class GoogleStorageResourceBucket implements WritableResource {
+
+	private final Storage storage;
+
+	private final String bucketName;
+
+	public GoogleStorageResourceBucket(Storage storage, String bucketName,
+			boolean createBucketIfNotExists) {
+		Assert.notNull(storage, "Storage object can not be null");
+		Assert.notNull(bucketName, "Bucket name can not be null");
+
+		this.storage = storage;
+		this.bucketName = bucketName;
+
+		if (!exists() && createBucketIfNotExists) {
+			this.storage.create(BucketInfo.newBuilder(this.bucketName).build());
+		}
+	}
+
+	@Override
+	public boolean exists() {
+		try {
+			return throwExceptionForNullBucket(resolve()).exists();
+		}
+		catch (IOException e) {
+			return false;
+		}
+	}
+
+	@Override
+	public boolean isReadable() {
+		return false;
+	}
+
+	@Override
+	public boolean isOpen() {
+		return false;
+	}
+
+	@Override
+	public URL getURL() throws IOException {
+		return getURI().toURL();
+	}
+
+	@Override
+	public URI getURI() throws IOException {
+		URI uri = null;
+		try {
+			uri = new URI(GoogleStorageProtocolResolver.PROTOCOL + getFilename());
+		}
+		catch (URISyntaxException e) {
+			throw new IOException("Invalid URI syntax", e);
+		}
+		return uri;
+	}
+
+	@Override
+	public File getFile() throws IOException {
+		throw new UnsupportedOperationException(
+				"Cannot get a file from a storage bucket.");
+	}
+
+	@Override
+	public long contentLength() throws IOException {
+		throw new UnsupportedOperationException(
+				"Cannot get content length for a storage bucket.");
+	}
+
+	@Override
+	public long lastModified() throws IOException {
+		throw new UnsupportedOperationException(
+				"Cannot get last-modified for a storage bucket.");
+	}
+
+	@Override
+	public Resource createRelative(String relativePath) throws IOException {
+		return new GoogleStorageResourceObject(this.storage,
+				getURI().toString() + (relativePath.startsWith("/") ? "" : "/")
```

Comment:
What if the bucket URI already ends in a `/`?
We'd want to avoid cases like `gs://bucket//blob`, but we should also take this chance to fix the case where we can't create blobs on buckets that don't end in a `/`.

Reply:
We don't store the entire bucket path in this class, just the bucket name. The reason is the GCS bucket functions all work on the bucket name, not the path. As a result, the path we construct here for the bucket will always not have the /.

Reply:
Also, i've got a unit test for this case , too.

Reply:
Yeah, I meant to ask about that -- do we really need them to work based on the name?
I'd rather have them work based on a `Bucket`, with `getName()` etc., unless there is a good reason not to.

Reply:
I actually think it's pretty clean to just have the bucket name as a string, and I would prefer to not have a Bucket that might refer to a nonexistent bucket on GCS. As it is now a GSRB is an object that refers to a potentially nonexistent bucket on GCS, and any actual Bucket we're using exists on GCS.

Reply:
How is calling `resolve()` on every method that needs the bucket _cleaner_?

And again, no reason why the state of this Java object in memory should map to the state of the GCS entity it refers to IMO.

Reply:
Again, I think this is an opinion thing. I think it's conceptually much cleaner, and calling resolve 3 or 4 times in the class isn't messy in my opinion.

Reply:
I'll defer to what others think. @artembilan @meltsufin

Reply:
See my other comment on `getFilename` which doesn't require a call to `resolve` and it also guarantees that there is no trailing `/`. So, as long as `getFilename` is fixed, simply using `this.bucketName`, this code is fine.